server/types/serialization.go

```diff
@@ -15,197 +15,152 @@
-	// that cannot be normally serialized. This is okay since these types cannot be deserialized, preventing them from
-	// being used outside of comparisons.
-	switch extendedType.(type) {
-	case AnyArrayType:
-		return []byte{0}
-	case UnknownType:
-		return []byte{1}
-	}
-	serializedType, err := SerializeType(extendedType)
-	if err != nil {
-		panic(err)
-	}
-	return serializedType
-}
-
 // DeserializeType is able to deserialize the given serialized type into an appropriate extended type. All extended
 // types will be defined by DoltgreSQL.
 func DeserializeType(serializedType []byte) (types.ExtendedType, error) {
-	if len(serializedType) < serializationIDHeaderSize {
-		return nil, fmt.Errorf("cannot deserialize an empty type")
+	return Deserialize(serializedType)
+}
+
+// Serialize returns the DoltgresType as a byte slice.
+func (t DoltgresType) Serialize() []byte {
+	writer := utils.NewWriter(256)
+	writer.VariableUint(0) // Version
+	// Write the type to the writer
+	writer.Uint32(t.OID)
+	writer.String(t.Name)
+	writer.String(t.Schema)
+	writer.String(t.Owner)
+	writer.Int16(t.TypLength)
+	writer.Bool(t.PassedByVal)
+	writer.String(string(t.TypType))
+	writer.String(string(t.TypCategory))
+	writer.Bool(t.IsPreferred)
+	writer.Bool(t.IsDefined)
+	writer.String(t.Delimiter)
+	writer.Uint32(t.RelID)
+	writer.String(t.SubscriptFunc)
+	writer.Uint32(t.Elem)
+	writer.Uint32(t.Array)
+	writer.String(t.InputFunc)
+	writer.String(t.OutputFunc)
+	writer.String(t.ReceiveFunc)
+	writer.String(t.SendFunc)
+	writer.String(t.ModInFunc)
+	writer.String(t.ModOutFunc)
+	writer.String(t.AnalyzeFunc)
+	writer.String(string(t.Align))
+	writer.String(string(t.Storage))
+	writer.Bool(t.NotNull)
+	writer.Uint32(t.BaseTypeOID)
+	writer.Int32(t.TypMod)
+	writer.Int32(t.NDims)
+	writer.Uint32(t.TypCollation)
+	writer.String(t.DefaulBin)
+	writer.String(t.Default)
+	writer.VariableUint(uint64(len(t.Acl)))
+	for _, ac := range t.Acl {
+		writer.String(ac)
 	}
-	serializationID, version := SerializationIDFromBytes(serializedType)
-	targetType, ok := serializationIDToType[serializationID]
-	if !ok {
-		return nil, fmt.Errorf("serialization ID %d does not have a matching type for deserialization", serializationID)
+	writer.VariableUint(uint64(len(t.Checks)))
+	for _, check := range t.Checks {
+		writer.String(check.Name)
+		writer.String(check.CheckExpression)
 	}
-	return targetType.deserializeType(version, serializedType[serializationIDHeaderSize:])
+	writer.Int32(t.AttTypMod)
+	// TODO: get rid this?
+	writer.String(t.internalName)
+	return writer.Data()
 }
 
-// serializationIDHeaderSize is the size of the header that applies to all serialization IDs.
-const serializationIDHeaderSize = 4
+// Deserialize returns the Collection that was serialized in the byte slice.
+// Returns an empty Collection if data is nil or empty.
+func Deserialize(data []byte) (DoltgresType, error) {
```

Comment:
We could move this code into `DeserializeType`, and do a forced cast to `DoltgresType` if we need to call that from other places.